### PR TITLE
ROC-1458: Implement nightly ZAP job

### PR DESCRIPTION
**Pull request by rlewan**
### JIRA link (if applicable) ###

https://tools.hmcts.net/jira/browse/ROC-1458

### Change description ###

Adds an overnight ZAP security scan. Executes using integration tests against a dockerized environment.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -6,8 +6,34 @@ This is the security scan using ZAP proxy.
...
 
+The following software needs to be installed:
+
 * [Docker](https://www.docker.com)
 
+### Local environment setup
+
+In addition to above, a link to integration tests should exist as security scan runs integration tests through ZAP proxy. Link can be created using the following command:
+
+```bash
```
Comment (timja) [inline README.md:17]: this isn't required on jenkins though right?

File: Jenkinsfile [hunk 2]
```diff
@@ -0,0 +1,83 @@
+#!groovy
+
+java.lang.String dockerCompose = 'docker-compose -f integration-tests/docker-compose.yml -f docker-compose.yml --project-directory .'
```
Comment (timja) [inline Jenkinsfile]: why not `String` ?

File: README.md (hunk 1 above)
Comment (rlewan) [reply README.md:17]: It isn't because Jenkins does a checkout to that directory so it has a local copy.

File: docker-compose.yml [hunk 3]
```diff
@@ -0,0 +1,20 @@
+version: '2.1'
+
+services:
+  zap-proxy:
+    image: owasp/zap2docker-weekly:latest
+    command: zap-x.sh -daemon -host 0.0.0.0 -config view.mode=attack -config scanner.attackOnStart=true -config api.disablekey=true -config api.addrs.addr.name=.* -config api.addrs.addr.regex=true -config connection.timeoutInSecs=120 -config database.recoverylog=false -config connection.proxyChain.hostName=reformmgmtproxyout.reform.hmcts.net -config connection.proxyChain.port=8080
```
Comment (rlewan) [inline docker-compose.yml]: I had to configure `reformmgmtproxyout` here as zap `-config` entries.

Comment (timja) [reply docker-compose.yml]: `proxyout.reform.hmcts.net` should be used

Comment (timja) [reply docker-compose.yml]: why does it need to talk out btw?

Comment (rlewan) [reply docker-compose.yml]: Payments

File: Jenkinsfile [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+  ],
+  [$class: 'VaultSecret', path: 'secret/dev/cmc/notify/integration-tests/test_mode_api_key', secretValues:
+    [
+      [$class: 'VaultSecretValue', envVar: 'GOV_NOTIFY_API_KEY', vaultKey: 'value']
+    ]
+  ]
+]
+
+String dockerCompose = 'docker-compose -f integration-tests/docker-compose.yml -f docker-compose.yml --project-directory .'
+String exec = 'exec -u `id -u` -T'
```
Comment (timja) [inline Jenkinsfile]: The $() syntax is better to use in general
```
String exec = 'exec -u $(id -u) -T
```

**Review by pragnesh-patel — COMMENTED** [review]
Looks good, will need a demo of this

**Review by pragnesh-patel — APPROVED** [review]
Where do you set the time for the Run overnight?

File: Jenkinsfile [hunk 5]
```diff
@@ -0,0 +1,84 @@
+#!groovy
+
+properties(
+  [[$class: 'GithubProjectProperty', displayName: 'Security scan', projectUrlStr: 'https://git.reform.hmcts.net/cmc/security-scan/'],
```
Comment (timja) [inline Jenkinsfile]: old url here

**Comment by rlewan** [comment]
I removed active scanning script and pipeline stage as they were redundant. ZAP will perform active scanning on each page it opens when started up in attack mode, which we do: https://github.com/zaproxy/zap-core-help/wiki/HelpStartConceptsModes

**Review by timja — APPROVED** [review]

**Pull request merged**